Add SSL options to launch_server

k4rth33k opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

Adding support for SSL certs so that sglang can be directly used with https endpoints. 

## Modifications

- Modify CLI args to take in arguments related to SSL certificates and pass them to uvicorn. 
- Update documentation for the newly added args

Benchmarks and tests were not added as this change is only for the server and not the sglang engine.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [x] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.


### How to run
Steps to test this feature:

1. Generate a self-signed certificate and key (Only if you do not have a CA signed certificate and key available):

```
openssl req -x509 -nodes -newkey rsa:2048 \
  -keyout test-sglang.com.key \
  -out test-sglang.com.cert \
  -days 365 \
  -subj "/CN=test-sglang.com"
```

2. Map test-sglang.com to 127.0.0.1 in /etc/hosts (This is so that the client can resolve the hostname to an IP address):
Note: If you are using Windows, you need to add a similar entry in C:\Windows\System32\drivers\etc\hosts

```
echo "127.0.0.1 test-sglang.com" | sudo tee -a /etc/hosts
```

3. Run the server:

Note: The `--ssl-self-signed-cert` flag should be passed only if you have a self-signed certificate and key.
```
python3 -m sglang.launch_server \
 --model-path /qwen_2_tiny \
 --host test-sglang.com \
 --max-total-tokens 512 \
 --ssl-certfile test-sglang.com.cert \
 --ssl-keyfile test-sglang.com.key \
 --ssl-self-signed-cert
```

Logs: 
```
...
[2025-04-06 20:38:30 TP0] max_total_num_tokens=512, chunked_prefill_size=2048, max_prefill_tokens=16384, max_running_requests=256, context_len=131072
[2025-04-06 20:38:31] INFO:     Started server process [2588]
[2025-04-06 20:38:31] INFO:     Waiting for application startup.
[2025-04-06 20:38:32] INFO:     Application startup complete.
[2025-04-06 20:38:32] INFO:     Uvicorn running on https://test-sglang.com:30000 (Press CTRL+C to quit)
[2025-04-06 20:38:33] INFO:     127.0.0.1:46014 - "GET /get_model_info HTTP/1.1" 200 OK
[2025-04-06 20:38:33 TP0] Prefill batch. #new-seq: 1, #new-token: 6, #cached-token: 0, token usage: 0.00, #running-req: 0, #queue-req: 0, 
[2025-04-06 20:38:34] INFO:     127.0.0.1:46014 - "POST /generate HTTP/1.1" 200 OK
[2025-04-06 20:38:34] The server is fired up and ready to roll!
```

### Use unitest
```bash
cd test/srt
python3 -m unittest test_srt_launch_server_ssl.TestSRTLaunchServerSSL
```

k4rth33k committed:
Add SSL options to launch_server and update docs

k4rth33k committed:
Formatting

k4rth33k commented:
Hi sglang team
Can anyone review this PR:
@zhyncs @ByronHsu @Ying1123

zhaochenyang20 committed:
Merge branch 'main' into k4rth33k/ssl

ryang-max commented:
Are there any test results about this feature? Could you please add the result to PR description? like launch server succeeded and requests can be fulfilled; compatibility to HTTP; compatibility to router; etc. And I suggest to add all SSL-related options in uvicorn here.

zhaochenyang20 commented:
> Are there any test results about this feature? Could you please add the result to PR description? like launch server succeeded and requests can be fulfilled; compatibility to HTTP; compatibility to router; etc. And I suggest to add all SSL-related options in uvicorn here.

@ryang-max thanks ruiguo!

@k4rth33k could you resolve the issue?

k4rth33k committed:
Use ssl cert in _wait_and_warmupl; Add new arg for self-signed certs

k4rth33k commented:
@ryang-max @zhaochenyang20 Thanks for the suggestions. `sglang.launch_server` is now working and the server is able to finish warmup and accept requests. I've updated the command used and result in PR description. 

Coming to your other questions, uvicorn doesn't support dual stack so it cannot handle both HTTP and HTTP requests at once. But I don't think it will be a problem. 
Reg. SSL args related to uvicorn, currently our team only requires certfile and keyfile since we are using a CA signed certificate. I've added the other args to match the options available in vLLM as of now. 

Thanks for pointing out sglang_router. I'll check if there are changes needed there.

zhaochenyang20 commented:
> @ryang-max @zhaochenyang20 Thanks for the suggestions. `sglang.launch_server` is now working and the server is able to finish warmup and accept requests. I've updated the command used and result in PR description.
> 
> Coming to your other questions, uvicorn doesn't support dual stack so it cannot handle both HTTP and HTTP requests at once. But I don't think it will be a problem. Reg. SSL args related to uvicorn, currently our team only requires certfile and keyfile since we are using a CA signed certificate. I've added the other args to match the options available in vLLM as of now.
> 
> Thanks for pointing out sglang_router. I'll check if there are changes needed there.

do we have any unit tests?

k4rth33k commented:
> > @ryang-max @zhaochenyang20 Thanks for the suggestions. `sglang.launch_server` is now working and the server is able to finish warmup and accept requests. I've updated the command used and result in PR description.
> > Coming to your other questions, uvicorn doesn't support dual stack so it cannot handle both HTTP and HTTP requests at once. But I don't think it will be a problem. Reg. SSL args related to uvicorn, currently our team only requires certfile and keyfile since we are using a CA signed certificate. I've added the other args to match the options available in vLLM as of now.
> > Thanks for pointing out sglang_router. I'll check if there are changes needed there.
> 
> do we have any unit tests?

@zhaochenyang20 Didn't find any existing unit tests for `sglang.srt.entrypoints.http_server.launch_server`. Where do you suggest to add a test for this?

zhaochenyang20 commented:
> > > @ryang-max @zhaochenyang20 Thanks for the suggestions. `sglang.launch_server` is now working and the server is able to finish warmup and accept requests. I've updated the command used and result in PR description.
> > > Coming to your other questions, uvicorn doesn't support dual stack so it cannot handle both HTTP and HTTP requests at once. But I don't think it will be a problem. Reg. SSL args related to uvicorn, currently our team only requires certfile and keyfile since we are using a CA signed certificate. I've added the other args to match the options available in vLLM as of now.
> > > Thanks for pointing out sglang_router. I'll check if there are changes needed there.
> > 
> > 
> > do we have any unit tests?
> 
> @zhaochenyang20 Didn't find any existing unit tests for `sglang.srt.entrypoints.http_server.launch_server`. Where do you suggest to add a test for this?

add a new test file and register the file in `run_suite.py`.

k4rth33k committed:
Merge branch 'main' into k4rth33k/ssl

k4rth33k committed:
Add unit test for launch_server with SSL args

k4rth33k committed:
Merge branch 'main' into k4rth33k/ssl

k4rth33k commented:
Hey @zhaochenyang20 
Bumping this PR. Unit test has been added. 
Right now, we are not using `sglang_router` in our deployments. Should SSL support for router be a different PR?

merrymercy committed:
Merge branch 'main' into k4rth33k/ssl

k4rth33k committed:
Merge branch 'main' into k4rth33k/ssl

k4rth33k committed:
Merge branch 'main' into k4rth33k/ssl

k4rth33k committed:
Merge branch 'main' into k4rth33k/ssl

k4rth33k commented:
@ByronHsu @zhaochenyang20 @merrymercy 
Can anyone review this PR please. 
Thanks.

k4rth33k committed:
Merge branch 'main' into k4rth33k/ssl

k4rth33k committed:
Merge branch 'main' into k4rth33k/ssl

zhaochenyang20 commented:
Please leave it here. We will find time to review. Thanks
